docs/proposals/11-apiserver-network-proxy.md

```diff
@@ -0,0 +1,134 @@
+
+#### Possible Solutions
+
+There are currently two ways to solve this problem:
+
+- Attach a port-forwarder side-car to prometheus.
+- Utilize the proxy subresource on the API server.
+
+#### Port-forwarder Sidecar
+
+With this solution each prometheus instance would have a side-car that has the kubeconfig of the shoot cluster, and which estalishes a port-forward connection to the endpoints residing in the shoot.
+
+There are a couple of problems with this approach:
+
+- the port-forward connection is not reliable.
+- the connection would break if the API server instance dies.
+- requires an additional component.
+
+#### Proxy sub-resource
+
+Kubernetes supports proxying requests to nodes, services, and pod endpoints in the shoot cluster. This proxy connection can be utilized for scraping the necessary endpoints in the shoot.
+
+This approach requires less-compenents and is more reliable than the port-forward solution, however, it relies on having the API server supporting proxied connection for the required endpoints.
+
+### Proxy-server Loadbalancer Sharing and Re-advertising
+
+With the proxy-server in place, we need to provide means to enable the proxy-agent in the shoot to establish the connection with the server. As a result, we need to provide a public endpoint through which this channel of communication can be established, i.e., we need a Loadbalancer(s).
+
+
+#### Possible Solution
+
+Using a Loadbalancer / proxy server would not make sense since this is a pain-point we are trying to eliminate in the first-place, doing so just moves the costs to the control-plane. A possible solution is to utilize proxy-protocol to communicate over a shared loadbalancer in the seed, similar to what has been proposed [here](https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md), this way we can prevent the extra-costs for load-balancers.
+
+With this in mind, we still have other pain-points, namely:
+
+- Advertising Loadbalancer public IPs to the shoot.
+- Directing the traffic to the corresponding shoot proxy-server.
+
+For advertising the Loadbalancer IP, a DNS entry can be created for the proxy loadbalancer, along with necessary certificates, which is then used to connect to the loadbalancer. At this point we can decided on either one of the two approaches:
+
+1. One Proxy / API server with a shared loadbalancer.
+2. Use one proxy server for all agents.
+
+In the first case, we will probably need a proxy for the proxy-server that knows how to direct traffic to the correct proxy server based on the corresponding shoot cluster. In the second case, we don't need another proxy if the proxy server is cluster-aware, i.e., can pool and identify connections coming from the same cluster and peer them with the correct API. Unfortunately, the second case is not supported today.
```

Comment:
I would suggest to do a solution very similar to what [API server SNI](https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md) is trying to achieve:

```
Shoot-A Proxy-Agent -> LB (shared) -> SNI (proxy) -> Shoot-A Proxy-Server
Shoot-B Proxy-Agent -> LB (shared) -> SNI (proxy) -> Shoot-B Proxy-Server
```

When combined with API server SNI, then you'll end with only 1 LB needed for the Seed cluster.
Even better - the API server DNS entry can be reused, but have a different port:

- `Shoot-A Proxy-Agent` is configured to talk to `api.<shoot-a>.<project>.internal.<domain>:9443`
- `Shoot-B Proxy-Agent` is configured to talk to `api.<shoot-b>.<project>.internal.<domain>:9443`

The SNI proxy would know that `api.<shoot-a>.<project>.internal.<domain>:9443` points to `Shoot-A Proxy-Server` and would simply forward the traffic to it.
There won't be a need for any `PROXY protocol` in this case.

Reply:
yup thats the first approach, but what you are  saying is even share the loadbalancer of the API-server and  increase the replicas by adding other services if required. I think that should work, we just need to know the loadbalancer limits.